_upgrading/sopel-8.md

```diff
@@ -0,0 +1,420 @@
+combining status information from across the bot's subsystems to answer the very
+important question, "Is the `bot` actually registered to an IRC network?" That
+is, Sopel not only has a _connection_ open, but the IRC server has _accepted_
+the bot as a _client_ and it's possible to _send commands_ to the IRC server.
+
+This is useful to plugins with code that runs without a triggering IRC event but
+still outputs to IRC, such as a `@sopel.plugin.interval()` job or acting on data
+received [via a socket][sopel-sockmsg]. `if not bot.connection_registered`, the
+output can be skipped, retried after a delay, etc.
+
+_Hopefully no one was overwriting the old simple attribute in plugin code, but
+now they **can't** do that._
+
+  [sopel-sockmsg]: https://github.com/dgw/sopel-sockmsg
+
+### More consistent `trigger` objects
+
+#### `STATUSMSG` handling
+
+Events sent to a channel can be scoped to users with a particular privilege
+level, or _status_, in that channel. IRC servers advertise the availability of
+this feature using the `STATUSMSG` token in `RPL_ISUPPORT`.
+
+In Sopel 7, `trigger.sender` included the status prefix, and it was difficult
+for plugins to detect and remove it themselves if so desired. Sopel 8.0 removes
+the status prefix from `trigger.sender` and leaves only the channel name, while
+the status prefix (if any) is exposed as `trigger.status_prefix`. This change is
+useful for plugins that use `trigger.sender` to store or retrieve _channel
+values_ in the bot's database, for example.
+
+The `bot` passed to a plugin callable triggered by an IRC event automatically
+includes the status prefix, if present, in the default `destination` parameter
+to methods that send a _message_ to IRC. Plugin callables that simply invoke
+`bot.say()`, `bot.reply()`, etc. in direct response to a `trigger` and without
+overriding the default `destination` will most likely get the expected behavior.
+
+[`sopel-remind`][sopel-remind] is an example of a plugin that this change might
+surprise, since the naïve implementation of such a reminder plugin would simply
+store `(trigger.nick, trigger.sender, parsed_message)` and send
+`bot.reply(parsed_message, trigger.sender, trigger.nick)` after the specified
+amount of time has passed. The `trigger.status_prefix` is lost, and the reminder
+sent to the entire channel, rather than the status-limited subset of users who
+could have seen the original command.
+
+#### Only set the `sender` when it makes sense
+
+For general IRC events like `QUIT`, `RPL_NAMREPLY`, etc. that don't happen "in"
+a channel or other clearly defined _context_, the `trigger.sender` value that
+plugin callables saw in Sopel 7 was unpredictable at best.
+
+Sopel 8.0 now _only_ sets the `trigger.sender` if the triggering event warrants
+it. In all other cases, the `sender` will be `None` and plugin code should use
+the `trigger.args` list to retrieve all information about the event.
```

Comment:
Sharing a rewrite of this section I did in my initial review. Main change is that it now lists out the events that do set `trigger.sender`

```suggestion
Sopel 8.0 now _only_ sets `trigger.sender` for the following events:
  * `INVITE`
  * `JOIN`
  * `KICK`
  * `MODE`
  * `NOTICE`
  * `PART`
  * `PRIVMSG`
  * `TOPIC`

In all other cases, the `sender` will be `None` and plugin code should use the
`trigger.args` list to retrieve all information about the event.

For general IRC events like `QUIT`, `RPL_NAMREPLY`, etc. that don't happen "in"
a channel or other clearly defined _context_, the `trigger.sender` value that
plugin callables saw in Sopel 7 was unpredictable at best.
```

Reply:
Once again, using your suggestion as a starting point for further polishing:

```suggestion
Sopel 8.0 now _only_ sets `trigger.sender` for the following events:

  * `INVITE`
  * `JOIN`
  * `KICK`
  * `MODE`
  * `NOTICE`
  * `PART`
  * `PRIVMSG`
  * `TOPIC`

In all other cases, the `sender` will be `None` and plugin code should use the
`trigger.args` list to retrieve all information about the event.

For general IRC events like `QUIT`, `RPL_NAMREPLY`, etc. that don't happen "in"
a channel or other clearly defined _context_, the `trigger.sender` value that
plugin callables saw in Sopel 7 was unpredictable at best. This change is
intended to make it easier to write correct behavior going forward.
```